coderd/schedule/template.go

```diff
@@ -5,32 +5,123 @@ import (
 	"time"
 
 	"github.com/google/uuid"
+	"golang.org/x/xerrors"
 
 	"github.com/coder/coder/coderd/database"
 )
 
+const MaxTemplateRestartRequirementWeeks = 16
+
+func TemplateRestartRequirementEpoch(loc *time.Location) time.Time {
+	// The "first week" starts on January 2nd, 2023, which is the first Monday
+	// of 2023. All other weeks are counted using modulo arithmetic from that
+	// date.
+	return time.Date(2023, time.January, 2, 0, 0, 0, 0, loc)
+}
+
+// DaysOfWeek intentionally starts on Monday as opposed to Sunday so the weekend
+// days are contiguous in the bitmap. This matters greatly when doing restarts
+// every second week or more to avoid workspaces restarting "at the start" of
+// the week rather than "at the end" of the week.
+var DaysOfWeek = []time.Weekday{
+	time.Monday,
+	time.Tuesday,
+	time.Wednesday,
+	time.Thursday,
+	time.Friday,
+	time.Saturday,
+	time.Sunday,
+}
+
+type TemplateRestartRequirement struct {
+	// DaysOfWeek is a bitmap of which days of the week the workspace must be
+	// restarted. If fully zero, the workspace is not required to be restarted
+	// ever.
+	//
+	// First bit is Monday, ..., seventh bit is Sunday, eighth bit is unused.
+	DaysOfWeek uint8
+	// Weeks is the amount of weeks between restarts. If 0 or 1, the workspace
+	// is restarted weekly in accordance with DaysOfWeek. If 2, the workspace is
+	// restarted every other week. And so forth.
+	//
+	// The limit for this value is 16, which is roughly 4 months.
+	//
+	// The "first week" starts on January 2nd, 2023, which is the first Monday
+	// of 2023. All other weeks are counted using modulo arithmetic from that
+	// date.
+	Weeks int64
+}
+
+// Days returns the days of the week that the workspace must be restarted.
+func (r TemplateRestartRequirement) Days() []time.Weekday {
+	days := make([]time.Weekday, 0, 7)
+	for i, day := range DaysOfWeek {
+		if r.DaysOfWeek&(1<<uint(i)) != 0 {
+			days = append(days, day)
+		}
+	}
+	return days
+}
+
+// DaysMap returns a map of the days of the week that the workspace must be
+// restarted.
+func (r TemplateRestartRequirement) DaysMap() map[time.Weekday]bool {
+	days := make(map[time.Weekday]bool)
+	for i, day := range DaysOfWeek {
+		days[day] = r.DaysOfWeek&(1<<uint(i)) != 0
```

Comment:
Should we make `r.DaysOfWeek&(1<<uint(i)) != 0` some helper func on it so `DaysMap` and `Days` use the same boolean func to decide to add it to their set?

Reply:
We're not using Days() so I removed it